README.md
# isomorph [![build status](https://secure.travis-ci.org/insin/isomorph.png)](http://travis-ci.org/insin/isomorph)

Utilities extracted from my dual-sided projects, which can be shared between
browsers and Node.js/io.js. This project is really a grab-bag, so modules are
intended to be required individually as needed, e.g. for type-checking functions
you would `require('isomorph/is')`.

## Install

```
npm install isomorph
```

Browser bundles export a global `isomorph` variable.

* [isomorph.js](https://raw.github.com/insin/isomorph/master/isomorph.js)
* [isomorph.min.js](https://raw.github.com/insin/isomorph/master/isomorph.min.js)

## Module APIs

## `is`

Type-checking and more - anything which is a generic, reusable test which would
naturally have a function name starting with "is".

#### Type-checking functions

`is.Array`,
`is.Boolean`,
`is.Date`,
`is.Error`,
`is.Function`,
`is.Number`,
`is.Object`,
`is.RegExp`,
`is.String`

Determines if the given input is of the specified type.

#### Content-checking functions

`is.Empty`

Determines if the given Object has any enumerable properties.

## `array`

Utilities for working with Arrays.

#### `array.flatten(arr)`

Flattens the given Array in place (and returns it, for chaining).

## `func`

Utilities for working with Functions.

#### `func.bind(fn, context[, arg1, ...])`

Binds the given function to the given execution context (`this` when
the function is called) and partially applies any additional arguments given.

The following properties are available on the binding function:

`__func__` - the function which is bound.

`__context__` - the context to which the function is bound.

## `object`

Utilities for working with Objects.

#### `object.hasOwn(obj, propertyName)`

Wrapped version of 1Object.prototype.hasOwnProperty()` - use to avoid the
"hasOwnProperty trap", as described in
[An Object is not a Hash](http://www.devthought.com/2012/01/18/an-object-is-not-a-hash/).

#### `object.type(obj)`

Returns the type of an object as a lowercase string:

```javascript
object.type({}) // "object"
object.type([]) // "array"
```

#### `object.extend(dest[, src1, ...])`

The classic `extend` method - copies own properties from `src` arguments to
`dest`, returning `dest`.

Does nothing for any `src` arguments which are falsy, so it's safe to pass in
an options argument which is potentially `undefined`, e.g.:

```javascript
function quiz(kwargs) {
 kwargs = object.extend({answer: 42, question: 'Meaning?'}, kwargs)
 // ...
}
```

#### `object.inherits(childConstructor, parentConstructor)`

The classic `inherits` method - puts `parentConstructor`'s prototype in
`childConstructor`'s prototype chain, returning `childConstructor`.

#### `object.items(obj)`

Creates an Array of `[property, value]` pairs from an Object.

#### `object.fromItems(items)`

Creates an Object from an Array of `[property, value]` pairs.

```javascript
function sortedFieldObj(fieldObj) {
   var fields = object.items(fieldObj)
   fields.sort(function(a, b) {
     return a[1].creationCounter - b[1].creationCounter
   })
   return object.fromItems(fields)
}
```

#### `object.lookup(arr)`

Creates a lookup Object from an Array, coercing each item in the Array to String
and adding it to a lookup Object as a property whose value is `true`:

```javascript
var ALLOWED_TAGS = ['div', 'span', 'h1']
  , TAG_LOOKUP = object.lookup(ALLOWED_TAGS)

function elementify(tagName) {
   if (TAG_LOOKUP[tagName]) {
      console.log(tagName + ' is valid')
   }
   else {
      console.log(tagName + ' is not allowed')
   }
}
```

#### `object.get(obj, prop, default)`

If the object has an own property with the given name, returns its value,
otherwise returns the given default.

#### `object.pop(obj, prop, default)`

If the object has an own property with the given name, deletes the property from
the object and returns its value, otherwise returns the given default.

#### `object.setDefault(obj, prop, default)`

If the object has an own property with the given name, returns its value,
otherwise sets the given default as the property and returns it.

## `format`

Formatting utilities.

#### `format.format(str[, r1, ...])`

Replaces `"%s"` placeholders in the given string with positional arguments.

#### `format.formatArr(str, arr)`

Replaces `"%s"` placeholders in the given string with arguments passed as
an Array.

To output a literal `'%'`, escape percentage signs by doubling them up:

```javascript
format.format('%% Complete: %s%%', 95) // '% Complete: 95%'
```

#### `format.formatObj(str, obj)`

Replaces `"{varName}"` placeholders in the given string with same-named
properties from a given object.

To output a literal '`{varName}'`, double up the opening brace:

```javascript
format.formatObj('{{foo}={foo}, {{bar}={bar}', {foo: 1, bar: 2}) // '{foo}=1, {bar}=2'
```

#### `format.fileSize(bytes[, threshold])`

Formats a number of bytes as a file size with an appropriately-scaled unit. The
threshold argument determines the point at which the next unit up is used,
defaulting to 768.

File sizes are rounded to the second decimal point, with any trailing zeros
being stripped off and the decimal point also being omitted if all decimals
are zero.

```javascript
format.fileSize(768) // '768 bytes'
format.fileSize(769) // '0.75 kB
format.fileSize(123456789) // '117.74 MB'
```

## `re`

Regular Expression utilities.

#### `re.findAll(regex, str[, flags])`

Uses a regular expression (given as a String or a RegExp object) to
find and return matches in the given String, in the vein of Python's
[`re.findall`](http://docs.python.org/library/re.html#re.findall).

If a single group is present in the RegExp, a list of matches will be
returned. If more than one group is present, a list of lists of matches
will be returned.

If `regex` is specified as a String, the `flags` argument can be used
to specify the flags to be used when compiling the RegExp.

## `querystring`

Utilities for working with query strings.

#### `querystring.parse(str)`

Creates an Object from a query string, providing values for names which are
present more than once as an Array.

#### `querystring.stringify(obj)`

Creates a query string from an Object, expecting names with multiple values
to be specified as an Array.

## `copy`

Utilites for creating copies of objects. This implementation is from Oran
Looney's [Deep Copy in JavaScript](http://oranlooney.com/deep-copy-javascript/)
and exposes the same interface.

#### `copy.copy(obj)`

Creates a shallow copy of an object.

#### `copy.deepCopy(obj)`

Creates a deep copy of an object.

## `time`

Utilities for formatting and parsing times and dates.

#### Formatting Directives

The following formatting directives are supported by `time.strftime` and
`time.strptime`:

Directive | Meaning
--------- | -----------------------------------------------------
`%b`      | Locale's abbreviated month name
`%B`      | Locale's full month name
`%d`      | Day of the month as a decimal number [01,31]
`%H`      | Hour (24-hour clock) as a decimal number [00,23]
`%I`      | Hour (12-hour clock) as a decimal number [00,12]
`%m`      | Month as a decimal number [01,12]
`%M`      | Minute as a decimal number [00,59]
`%p`      | Locale's equivalent of either AM or PM (only with %I)
`%S`      | Second as a decimal number [00,59]
`%y`      | Year without century as a decimal number [00,99]
`%Y`      | Year with century as a decimal number
`%%`      | A literal `%` character

#### `time.strftime(date, format[, locale])`

A partial implementation of `strftime`, which formats a Date according to a
format string. An Error will be thrown if an invalid format string is given.

#### `time.strpdate(string, format[, locale])`

Parses time details from a string, based on a format string, returning a Date.

This is a convenience wrapper around `time.strptime`:

#### `time.strptime(string, format[, locale])`

A partial implementation of `strptime`, which parses time details from a
string, based on a format string.

Returns an Array of numbers, each corresponding to a datetime field:

Index | Represents | Values
----- | ---------- | ------------------
`0`   | Year       | (for example, 2003
`1`   | Month      | range [1,12]
`2`   | Day        | range [1,31]
`3`   | Hour       | range [0,23]
`4`   | Minute     | range [0,59]
`5`   | Second     | range [0,59]

This implementation largely takes its cue from the documentation for Python's
`time` module, as documented at http://docs.python.org/lib/module-time.html
with the exception of seconds formatting, which is restricted to the range
[00,59] rather than [00,61].

#### Locales

The `time` module has basic support for using locales when parsing and
formatting dates.

`time.defaultLocale` - the code for the default locale - defaults to `'en'`.

`time.locales` - an object defining locale details, with locale codes as its
properties. Only contains the locale definition for `'en'` by default.

`time.getLocale(code)` - retrieves the locale with the given code, falling back
to just the language code and finally to the default locale if a locale can't be
found.

Locale codes can consist of a language code (e.g. `'en'`) or a language
and region code (e.g. `'en-GB'`).

## `url`

Utilities for working with URLs.

#### `url.parseUri(url)`

Splits any well-formed URI into its parts - from
http://blog.stevenlevithan.com/archives/parseuri.

```
parseUri 1.2.2
(c) Steven Levithan <stevenlevithan.com>
MIT License
```

#### `url.makeUri(obj)`

Creates a URI from an object specification - from https://gist.github.com/1121696.

```
makeURI 1.2.2 - create a URI from an object specification
(c) Niall Smart <niallsmart.com>
MIT
```

### MIT Licensed

CHANGES.md
## v0.3.0 / 2014-11-11

* `copy.copy()` and `copy.deepCopy()` now handle primitive wrapper objects
  and copy over any custom properties added to them
* Added `object.type()`

## v0.2.1 / 2014-02-21

* Added `object.pop()`
* Added `object.setDefault()`

## v0.2.0 / 2012-06-29

* Moved modules up from `lib/` to `/` for tidier importing.

## v0.1.10 / 2012-06-27

* Added `format.fileSize()`

## v0.1.9 / 2012-03-08

* Extracted third-party functions from [newforms](https://github.com/insin/newforms)
  for reuse across projects:

  * `url.parseUri()`
  * `url.makeUri()`

## v0.1.8 / 2012-02-01

* Fixed `object.hasOwn()` in IE 7/8, where `Function.prototype.call.bind` is
  Not A Thing.

## v0.1.7 / 2012-01-28

* Added time module, with `time.strftime`, `time.strpdate`, `time.strftime`
  and some basic locale support.
* Added copy module, with `copy.copy()` and `copy.deepCopy()`
* Added `object.get()`

## v0.1.6 / 2012-01-25

* Added `object.hasOwn()`

## v0.1.5 / 2012-01-20

* Extracted from [sacrum](https://github.com/insin/sacrum):

  * `querystring.parse()`
  * `querystring.stringify()`

## v0.1.4 / 2012-01-19

* Extracted from [DOMBuilder](https://github.com/insin/DOMBuilder):

  * `array.flatten()`

## v0.1.3 / 2012-01-19

* Version number was incorrect in v0.1.2 build

## v0.1.2 / 2012-01-19

* Changed `function.bind()` now exposes the bound function as `__func__`
  and the bound context as `__context__`
* Extracted from [newforms](https://github.com/insin/newforms)

   * `object.items()` - `{a: 1, b: 2}` → `[['a', 1], ['b', 2]]`
   * `object.fromItems()` - `[['a', 1], ['b', 2]]` → `{a: 1, b : 2}`
   * `object.lookup()` - `['a', 'b', 'c']` → `{a: true, b: true: c: true}`

* Changed: `object.extend()` is now variadic, accepting multiple source
  objects
* Changed: escape `'%'` in `format.format()` and `format.formatArr()`
  using `'%%'`
* Changed: escape `'{varName}'` in `format.formatObj()` using
  `'{{varName}'`

## v0.1.1 / 2012-01-18

* Changed build script to use [buildumb](https://github.com/insin/buildumb)

## v0.1.0 / 2012-01-17

* Changed code structure - now written as a regular Node.js module
* Added a browser build script
* Changed API for browser use - now exposed as a global `isomorph` variable

## v0.0.2 / 2012-01-17

* Added `object.inherits()`

## v0.0.1 / 2011-12-15

* Initial release.


time.js
'use strict';

var is = require('./is')

/**
 * Pads a number with a leading zero if necessary.
 */
function pad(number) {
  return (number < 10 ? '0' + number : number)
}

/**
 * Returns the index of item in list, or -1 if it's not in list.
 */
function indexOf(item, list) {
  for (var i = 0, l = list.length; i < l; i++) {
    if (item === list[i]) {
      return i
    }
  }
  return -1
}

/**
 * Maps directive codes to regular expression patterns which will capture the
 * data the directive corresponds to, or in the case of locale-dependent
 * directives, a function which takes a locale and generates a regular
 * expression pattern.
 */
var parserDirectives = {
  // Locale's abbreviated month name
  'b': function(l) { return '(' + l.b.join('|') + ')' }
  // Locale's full month name
, 'B': function(l) { return '(' + l.B.join('|') + ')' }
  // Locale's equivalent of either AM or PM.
, 'p': function(l) { return '(' + l.AM + '|' + l.PM + ')' }
, 'd': '(\\d\\d?)' // Day of the month as a decimal number [01,31]
, 'H': '(\\d\\d?)' // Hour (24-hour clock) as a decimal number [00,23]
, 'I': '(\\d\\d?)' // Hour (12-hour clock) as a decimal number [01,12]
, 'm': '(\\d\\d?)' // Month as a decimal number [01,12]
, 'M': '(\\d\\d?)' // Minute as a decimal number [00,59]
, 'S': '(\\d\\d?)' // Second as a decimal number [00,59]
, 'y': '(\\d\\d?)' // Year without century as a decimal number [00,99]
, 'Y': '(\\d{4})'  // Year with century as a decimal number
, '%': '%'         // A literal '%' character
}

/**
 * Maps directive codes to functions which take the date to be formatted and
 * locale details (if required), returning an appropriate formatted value.
 */
var formatterDirectives = {
  'a': function(d, l) { return l.a[d.getDay()] }
, 'A': function(d, l) { return l.A[d.getDay()] }
, 'b': function(d, l) { return l.b[d.getMonth()] }
, 'B': function(d, l) { return l.B[d.getMonth()] }
, 'd': function(d) { return pad(d.getDate(), 2) }
, 'H': function(d) { return pad(d.getHours(), 2) }
, 'M': function(d) { return pad(d.getMinutes(), 2) }
, 'm': function(d) { return pad(d.getMonth() + 1, 2) }
, 'S': function(d) { return pad(d.getSeconds(), 2) }
, 'w': function(d) { return d.getDay() }
, 'Y': function(d) { return d.getFullYear() }
, '%': function(d) { return '%' }
}

/** Test for hanging percentage symbols. */
var strftimeFormatCheck = /[^%]%$/

/**
 * A partial implementation of strptime which parses time details from a string,
 * based on a format string.
 * @param {String} format
 * @param {Object} locale
 */
function TimeParser(format, locale) {
  this.format = format
  this.locale = locale
  var cachedPattern = TimeParser._cache[locale.name + '|' + format]
  if (cachedPattern !== undefined) {
    this.re = cachedPattern[0]
    this.matchOrder = cachedPattern[1]
  }
  else {
    this.compilePattern()
  }
}

/**
 * Caches RegExps and match orders generated per locale/format string combo.
 */
TimeParser._cache = {}

TimeParser.prototype.compilePattern = function() {
  // Normalise whitespace before further processing
  var format = this.format.split(/(?:\s|\t|\n)+/).join(' ')
    , pattern = []
    , matchOrder = []
    , c
    , directive

  for (var i = 0, l = format.length; i < l; i++) {
    c = format.charAt(i)
    if (c != '%') {
      if (c === ' ') {
        pattern.push(' +')
      }
      else {
        pattern.push(c)
      }
      continue
    }

    if (i == l - 1) {
      throw new Error('strptime format ends with raw %')
    }

    c = format.charAt(++i)
    directive = parserDirectives[c]
    if (directive === undefined) {
      throw new Error('strptime format contains an unknown directive: %' + c)
    }
    else if (is.Function(directive)) {
      pattern.push(directive(this.locale))
    }
    else {
      pattern.push(directive)
    }

    if (c != '%') {
       matchOrder.push(c)
    }
  }

  this.re = new RegExp('^' + pattern.join('') + '$')
  this.matchOrder = matchOrder
  TimeParser._cache[this.locale.name + '|' + this.format] = [this.re, matchOrder]
}

/**
 * Attempts to extract date and time details from the given input.
 * @param {string} input
 * @return {Array.<number>}
 */
TimeParser.prototype.parse = function(input) {
  var matches = this.re.exec(input)
  if (matches === null) {
    throw new Error('Time data did not match format: data=' + input +
                    ', format=' + this.format)
  }

    // Default values for when more accurate values cannot be inferred
  var time = [1900, 1, 1, 0, 0, 0]
    // Matched time data, keyed by directive code
    , data = {}

  for (var i = 1, l = matches.length; i < l; i++) {
    data[this.matchOrder[i - 1]] = matches[i]
  }

  // Extract year
  if (data.hasOwnProperty('Y')) {
    time[0] = parseInt(data.Y, 10)
  }
  else if (data.hasOwnProperty('y')) {
    var year = parseInt(data.y, 10)
    if (year < 68) {
        year = 2000 + year
    }
    else if (year < 100) {
        year = 1900 + year
    }
    time[0] = year
  }

  // Extract month
  if (data.hasOwnProperty('m')) {
    var month = parseInt(data.m, 10)
    if (month < 1 || month > 12) {
      throw new Error('Month is out of range: ' + month)
    }
    time[1] = month
  }
  else if (data.hasOwnProperty('B')) {
    time[1] = indexOf(data.B, this.locale.B) + 1
  }
  else if (data.hasOwnProperty('b')) {
    time[1] = indexOf(data.b, this.locale.b) + 1
  }

  // Extract day of month
  if (data.hasOwnProperty('d')) {
    var day = parseInt(data.d, 10)
    if (day < 1 || day > 31) {
      throw new Error('Day is out of range: ' + day)
    }
    time[2] = day
  }

  // Extract hour
  var hour
  if (data.hasOwnProperty('H')) {
    hour = parseInt(data.H, 10)
    if (hour > 23) {
      throw new Error('Hour is out of range: ' + hour)
    }
    time[3] = hour
  }
  else if (data.hasOwnProperty('I')) {
    hour = parseInt(data.I, 10)
    if (hour < 1 || hour > 12) {
      throw new Error('Hour is out of range: ' + hour)
    }

    // If we don't get any more information, we'll assume this time is
    // a.m. - 12 a.m. is midnight.
    if (hour == 12) {
        hour = 0
    }

    time[3] = hour

    if (data.hasOwnProperty('p')) {
      if (data.p == this.locale.PM) {
        // We've already handled the midnight special case, so it's
        // safe to bump the time by 12 hours without further checks.
        time[3] = time[3] + 12
      }
    }
  }

  // Extract minute
  if (data.hasOwnProperty('M')) {
    var minute = parseInt(data.M, 10)
    if (minute > 59) {
        throw new Error('Minute is out of range: ' + minute)
    }
    time[4] = minute
  }

  // Extract seconds
  if (data.hasOwnProperty('S')) {
    var second = parseInt(data.S, 10)
    if (second > 59) {
      throw new Error('Second is out of range: ' + second)
    }
    time[5] = second
  }

  // Validate day of month
  day = time[2], month = time[1], year = time[0]
  if (((month == 4 || month == 6 || month == 9 || month == 11) &&
      day > 30) ||
      (month == 2 && day > ((year % 4 === 0 && year % 100 !== 0 ||
                             year % 400 === 0) ? 29 : 28))) {
    throw new Error('Day is out of range: ' + day)
  }

  return time
}

var time  = {
  /** Default locale name. */
  defaultLocale: 'en'

  /** Locale details. */
, locales: {
    en: {
      name: 'en'
    , a: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
    , A: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday',
          'Friday', 'Saturday']
    , AM: 'AM'
    , b: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep',
          'Oct', 'Nov', 'Dec']
    , B: ['January', 'February', 'March', 'April', 'May', 'June', 'July',
          'August', 'September', 'October', 'November', 'December']
    , PM: 'PM'
    }
  }
}

/**
 * Retrieves the locale with the given code.
 * @param {string} code
 * @return {Object}
 */
var getLocale = time.getLocale = function(code) {
  if (code) {
    if (time.locales.hasOwnProperty(code)) {
      return time.locales[code]
    }
    else if (code.length > 2) {
      // If we appear to have more than a language code, try the
      // language code on its own.
      var languageCode = code.substring(0, 2)
      if (time.locales.hasOwnProperty(languageCode)) {
        return time.locales[languageCode]
      }
    }
  }
  return time.locales[time.defaultLocale]
}

/**
 * Parses time details from a string, based on a format string.
 * @param {string} input
 * @param {string} format
 * @param {string=} locale
 * @return {Array.<number>}
 */
var strptime = time.strptime = function(input, format, locale) {
  return new TimeParser(format, getLocale(locale)).parse(input)
}

/**
 * Convenience wrapper around time.strptime which returns a JavaScript Date.
 * @param {string} input
 * @param {string} format
 * @param {string=} locale
 * @return {date}
 */
time.strpdate = function(input, format, locale) {
  var t = strptime(input, format, locale)
  return new Date(t[0], t[1] - 1, t[2], t[3], t[4], t[5])
}

/**
 * A partial implementation of <code>strftime</code>, which formats a date
 * according to a format string. An Error will be thrown if an invalid
 * format string is given.
 * @param {date} date
 * @param {string} format
 * @param {string=} locale
 * @return {string}
 */
time.strftime = function(date, format, locale) {
  if (strftimeFormatCheck.test(format)) {
    throw new Error('strftime format ends with raw %')
  }
  locale = getLocale(locale)
  return format.replace(/(%.)/g, function(s, f) {
    var code = f.charAt(1)
    if (typeof formatterDirectives[code] == 'undefined') {
      throw new Error('strftime format contains an unknown directive: ' + f)
    }
    return formatterDirectives[code](date, locale)
  })
}

module.exports = time
